### ISSUE TITLE
Testing develop branch for 1.2.0 release

### ISSUE THREAD

**Issue by philippReist** [philippReist/DanceBotsEditor#80]
Hi @tiemes and @robinhanhart,

Looks like we are there with 1.2.0! Could you (or task someone from M&P, Tobias) please thoroughly test the develop branch before I merge it to master and we do the 1.2.0 release?

What I just did:
- Fix the umlaut issue in the song title/artist input fields
- Expand the howto text in the (?) button - please check!

Otherwise, please check that the new features work well with robots, and especially also on Mac which I have not tested yet.

New features for 1.2.0:
- Instaplay
- toggle for channel swap
- installer (to be added still, but is orthogonal to feature testing).

Once I get the OK from you we can proceed. Thank you for your help!
Philipp

**Comment by tiemes** [comment]
Hi Philipp
I'll get in contact with our team to have it tested.
I have someone who will provide us with a free to use mp3 file. However it will take some more time for him to create it.

**Comment by tiemes** [comment]
@philippReist do you have the binaries somewhere? Would make testing easier, we're quite busy with CYBATHLON projects.

**Comment by philippReist** [comment]
Not for Mac - I can make a zip for Windows. But we really need to test Mac as well.

**Comment by tiemes** [comment]
I‘d appreciate the windows zip. I tried to build it on my M1 mac, but had errors. Ill follow up next week.


Von meinem iPhone gesendet

**Comment by philippReist** [comment]
Here you go for the Win build: https://drive.google.com/file/d/1uw4eIWdBWZYrmUQzbM_FEDDhaB_5ttgd/view?usp=sharing

It would be great to test on Mac too before we release it.

**Comment by robinhanhart** [comment]
I tried testing it on my Intel Mac but I'm having cmake issues. Don't think they are related to Dancebots code.

`ld: can't map file, errno=22 file '/Library/Frameworks/sqlite3.framework' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)`

**Comment by tiemes** [comment]
[error.txt](https://github.com/philippReist/DanceBotsEditor/files/7202373/error.txt)

 94%] Linking CXX executable test-beatdetect
ld: warning: ignoring file /Users/toby/Qt/5.12.9/clang_64/lib/QtWidgets.framework/QtWidgets, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
ld: warning: ignoring file /Users/toby/Qt/5.12.9/clang_64/lib/QtGui.framework/QtGui, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
ld: warning: ignoring file /Users/toby/Qt/5.12.9/clang_64/lib/QtCore.framework/QtCore, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
  "QArrayData::deallocate(QArrayData*, unsigned long, unsigned long)", referenced from:
      QString::~QString() in main.cc.o
      operator+(QString const&, char const*) in main.cc.o
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
      QList<QString>::append(QString const&) in main.cc.o
      QString::toStdString() const in main.cc.o
      QList<QString>::detach_helper(int) in main.cc.o
      QList<QString>::detach_helper_grow(int, int) in main.cc.o
      ...
  "QArrayData::shared_null", referenced from:
      AudioFile::AudioFile() in audio_file.cc.o
      AudioFile::findHeaderCode(QFile*) in audio_file.cc.o
      AudioFile::clear() in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QByteArray::reallocData(unsigned int, QFlags<QArrayData::AllocationOption>)", referenced from:
      AudioFile::findHeaderCode(QFile*) in audio_file.cc.o
  "QByteArray::clear()", referenced from:
      AudioFile::clear() in audio_file.cc.o
  "QByteArray::append(char)", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QByteArray::resize(int)", referenced from:
      AudioFile::findHeaderCode(QFile*) in audio_file.cc.o
  "QByteArray::QByteArray(char const*, int)", referenced from:
      __GLOBAL__sub_I_audio_file.cc in audio_file.cc.o
  "QDataStream::setByteOrder(QDataStream::ByteOrder)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::applyDataStreamSettings(QDataStream*) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QDataStream::setFloatingPointPrecision(QDataStream::FloatingPointPrecision)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::applyDataStreamSettings(QDataStream*) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QDataStream::QDataStream(QByteArray*, QFlags<QIODevice::OpenModeFlag>)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QDataStream::QDataStream(QByteArray const&)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QDataStream::~QDataStream()", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QDataStream::operator<<(int)", referenced from:
      AudioFile::save(QString) in audio_file.cc.o
  "QDataStream::operator>>(int&)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QFileDevice::seek(long long)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QDir::QDir(QString const&)", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QDir::~QDir()", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QFile::open(QFlags<QIODevice::OpenModeFlag>)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QFile::QFile(QString const&)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QFile::~QFile()", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::save(QString) in audio_file.cc.o
  "QString::toUtf8_helper(QString const&)", referenced from:
      QString::toStdString() const in main.cc.o
      QString::toStdString() const in audio_file.cc.o
  "QString::fromUtf8_helper(char const*, int)", referenced from:
      operator+(QString const&, char const*) in main.cc.o
  "QString::fromAscii_helper(char const*, int)", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
      __GLOBAL__sub_I_main.cc in main.cc.o
  "QString::append(QString const&)", referenced from:
      operator+(QString const&, char const*) in main.cc.o
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QString::operator=(QString const&)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QIODevice::read(char*, long long)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
      AudioFile::findHeaderCode(QFile*) in audio_file.cc.o
  "QIODevice::read(long long)", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QIODevice::write(char const*, long long)", referenced from:
      AudioFile::save(QString) in audio_file.cc.o
  "QListData::detach_grow(int*, int)", referenced from:
      QList<QString>::detach_helper_grow(int, int) in main.cc.o
  "QListData::shared_null", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QListData::append()", referenced from:
      QList<QString>::append(QString const&) in main.cc.o
  "QListData::detach(int)", referenced from:
      QList<QString>::detach_helper(int) in main.cc.o
  "QListData::dispose(QListData::Data*)", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
      QList<QString>::detach_helper(int) in main.cc.o
      QList<QString>::detach_helper_grow(int, int) in main.cc.o
  "QListData::realloc(int)", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QFileDevice::pos() const", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QDir::exists() const", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QDir::filePath(QString const&) const", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QDir::entryList(QStringList const&, QFlags<QDir::Filter>, QFlags<QDir::SortFlag>) const", referenced from:
      (anonymous namespace)::BeatDetectTest_beatConsistency_Test::TestBody() in main.cc.o
  "QFile::size() const", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
  "QFile::exists() const", referenced from:
      AudioFile::load(QString) in audio_file.cc.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [test/test_beatdetect/test-beatdetect] Error 1
make[1]: *** [test/test_beatdetect/CMakeFiles/test-beatdetect.dir/all] Error 2
make: *** [all] Error 2

M1 Mac

**Comment by philippReist** [comment]
Hmm, @robinhanhart on an Intel Mac it should work with the instructions we have - can you please double-check that you followed them exactlz?

@tiemes - can you cross-compile for x86, i.e. Intel? Maybe you can choose the correct target/platform. Maybe in cmake? I think the intel apps should run on M1 if I am not mistaken.

**Comment by robinhanhart** [comment]
> Hmm, @robinhanhart on an Intel Mac it should work with the instructions we have - can you please double-check that you followed them exactlz?
> 
> 

I think I have a config problem that's specific to my work mac. Will check on my personal Intel mac when I get the chance.
Followed the instructions on the readme.

**Comment by philippReist** [comment]
@tiemes - right cross-compiling could be tricky. So much better to just use an Intel Macbook then. Do you still have the M&P Macbook I used a while back? The toolchain and everything should be installed on it.

@robinhanhart - thank you for checking with your personal Mac! I hope it'll work!

**Comment by tiemes** [comment]
I‘ll check.

**Comment by robinhanhart** [comment]
Also doesn't work on my personal mac. 😦 Same error. Maybe some issue with higher mac versions.
Couldn't find a working fix online.

macOS 11.6 Big Sur
Xcode 13.0

Error occurs on `[ 25%] Linking C executable sndfile-regtest`

```
ld: warning: directory not found for option '-L/opt/local/lib'
ld: can't map file, errno=22 file '/Library/Frameworks/sqlite3.framework' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libsndfile/sndfile-regtest] Error 1
make[1]: *** [lib/libsndfile/CMakeFiles/sndfile-regtest.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by r-oung** [comment]
Sorry for being late to the game.

@philippReist I might not have explained this to you before: 
- If you create a pull-request in the `develop` branch, it should automatically trigger the "GitHub build action".
- If you create a `Release`, it should automatically build a binary for macOS and upload to the release's assets. @robinhanhart There's a binary for macOS here (darwin): https://github.com/r-oung/DanceBotsEditor/releases/tag/v1.2.0-dev (forked repo for testing). The build seems to work correctly for macOS 11.0 + Qt 5.12.6 (https://github.com/r-oung/DanceBotsEditor/runs/3763338600?check_suite_focus=true), but there are issues with the deployment and release upload part of the script.

I tested the binary (see above) on macOS 10.15 (Catalina):

> * Fix the umlaut issue in the song title/artist input fields
- When I loaded an MP3 (with umlaut in the file's metadata title), it gave me the error: "Cannot decode corrupt MP3 file." However, this error might be unrelated.
- I am able to write umlauts into the DanceBots Editor Artist and Title fields, save the file, and reload the file correctly.

> * Expand the howto text in the (?) button - please check!
There's a typo. I'll submit a PR after this.

> New features for 1.2.0:
> 
> * Instaplay
This works correctly.
UI/UX suggestions:
- The message appears and disappears too quickly to know what's going on.
- Highlight the `Robot` and `Human` button when pressed so that it's clear which mode is being used.

> * toggle for channel swap
This works correctly.

Comment:
- When pressing `CLEAR`, everything is cleared except for the Artist and Title field. Consider clearing these as well.

**Comment by philippReist** [comment]
Whoa thank you so much Ray! @tiemes FYI you can just take that MacOS binary for testing / I think with Ray's testing we are good to publish this.

@robinhanhart would you be able to address the UI suggestions that Ray has? I agree that the highlight would be very useful.

I will take care of the CLEAR artist title text.

**Comment by philippReist** [comment]
> If you create a Release, it should automatically build a binary for macOS and upload to the release's assets

Is this also true for the develop branch, @r-oung?

**Comment by r-oung** [comment]
@philippReist Yes. If you create a release on any branch it should generate binaries. 

However, I realized today that there are some bugs with it. Don't tick the "This is a pre-release" checkbox, because that runs the build system twice for some reason. As a workaround, just do a standard release and tag with `-dev` or something, e.g. `v1.2.0-dev`.

**Comment by tiemes** [comment]
awesome - thanks! However I cannot access the links (404).

**Comment by r-oung** [comment]
@tiemes Sorry, I was making some changes and forgot that it would break the link.
This should work: https://github.com/r-oung/DanceBotsEditor/releases/tag/v1.2.0. Click on the DMG file.

**Comment by philippReist** [comment]
Thank you @r-oung!

**Comment by philippReist** [comment]
Ok I think we are ready for release.

- @r-oung fixed the typos - there's not enough space for the robot/human explanations so I left them out.
- I fixed the better highlights of whether HUMAN/ROBOT is active
- I added a delay so that the messages when switching between HUMAN / ROBOT are better readable. Kept them short at 250ms though.

I'll merge this to develop and then merge to master and create a new release.
